### PR TITLE
S3 data dump workflow updates

### PR DESCRIPTION
**Pull request by jvntf**
- [x] Modify stac-gen workflow to conditionally parse datetime from filenames
- [x] Add s3 discovery task 
- [x] new state machine to run / test it
- [x] document AWS permissions setting
- [x] document s3 discovery code

### CODE REVIEW

**Comment by jvntf** [comment]
psql > INSERT INTO pgstac.collections (content)
VALUES('{
   "id":"OMSO2PCA",
   "type": "Collection",
   "links":[
   ],
   "title":"OMSO2PCA",
   "extent":{
      "spatial":{
         "bbox":[
            [
               -180,
               -90,
               180,
               90
            ]
         ]
      },
      "temporal":{
         "interval":[
            [
               "2005-01-01T00:00:00Z",
               "2100-01-01T00:00:00Z"
            ]
         ]
      }
   },
   "license":"public-domain",
   "description":"OMI/Aura Sulphur Dioxide (SO2) Total Column 1-orbit L2 Swath 13x24 km V003 (OMSO2)",
   "stac_version":"1.0.0"
}');

**Comment by jvntf** [comment]
This is ready for review @abarciauskas-bgse.
A couple of questions / things to note:
- I have defined the second state machine within the same `cdk_stack.py` file that was already existing, which also required creating different tasks (only by name) that use the same lamdas. This all seemed quite verbose to me, but I wasn't sure what convention is with this-- I had tried breaking it up differently but ran into issues / seemed to be a misunderstanding of how the deploy was working so I rolled back.
- I inserted a collection into the database for `OMSO2PCA` containing the above metadata, just to have something there so i could test the rest of the workflow
- Getting this to work requires a whole bunch of AWS config, please let me know if this is clear enough from the docs I added, or if more info is needed

File: cogify/README.md [hunk 1]
```diff
@@ -40,3 +40,9 @@ python3 ERA5/fetch.py
 # Generate the cog
 python3 run.py -f download.nc -c ERA5
 ```
+
+
+AWS Provisioning:
+To function as a lambda task, the Lambda function's execution role needs to be given permission to AWS PutObject.
+
+- Add permission policy to the Lambda's execution role to allow `s3:PutObject`. This is because this Cogify lambda needs to access an S3 Bucket to write COG files
```
Comment (abarciauskas-bgse) [inline cogify/README.md:48]: I would guess that this permission policy is something we can add programmatically through CDK, but for now you can add an issue to this repo to do that.

File: cdk/cdk/cdk_stack.py [hunk 2]
```diff
@@ -69,44 +88,75 @@ def __init__(self, scope: core.Construct, construct_id: str, **kwargs) -> None:
...
-        discover_task = tasks.LambdaInvoke(
-            self, "Discover Granules Task", lambda_function=discover_lambda
+        ## CMR Workflow State Machine Steps
+        cmr_start_state = stepfunctions.Pass(self, "CMR Discovery StartState")
+        s3_start_state = stepfunctions.Pass(self, "S3 Discovery StartState")
+        cmr_discover_task = tasks.LambdaInvoke(
+            self, "CMR Discover Granules Task", lambda_function=cmr_discover_lambda
         )
+        s3_discover_task = tasks.LambdaInvoke(
+            self, "S3 Discover COGs Task", lambda_function=s3_discover_lambda
```
Comment (abarciauskas-bgse) [inline cdk/cdk/cdk_stack.py:98]: I think this S3 task should be generic to file type - so this should just be "S3 Discover Task"

File: cdk/cdk/cdk_stack.py [hunk 3]
```diff
@@ -69,44 +88,75 @@ def __init__(self, scope: core.Construct, construct_id: str, **kwargs) -> None:
...
 
-        simple_state_machine = stepfunctions.StateMachine(
-            self, f"{collection}-COG-StateMachine", definition=definition
+        s3_wflow_state_machine = stepfunctions.StateMachine(
+            self, f"{bucket}-{prefix}-COG-StateMachine", definition=s3_wflow_definition
         )
 
+
         # Rule to run it
         rule = events.Rule(
```
Comment (abarciauskas-bgse) [inline cdk/cdk/cdk_stack.py:154]: In general, I think these schedule rules will be per state machine. However in this particular case I don't think we are expecting new files on a regular cadence so I don't there there should be any rule for the discover S3 -> generate STAC workflow we are creating.

File: s3-discovery/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+RUN find . -type f -name '*.pyc' | while read f; do n=$(echo $f | sed 's/__pycache__\///' | sed 's/.cpython-[2-3][0-9]//'); cp $f $n; done;
+RUN find . -type d -a -name '__pycache__' -print0 | xargs -0 rm -rf
+RUN find . -type f -a -name '*.py' -print0 | xargs -0 rm -f
+RUN find . -type d -a -name 'tests' -print0 | xargs -0 rm -rf
+RUN echo "Remove lambda python packages"
+RUN rm -rdf ./numpy/doc/
+RUN rm -rdf ./stack
+RUN rm -rdf ./docutils*
+
+COPY handler.py handler.py
```
Comment (abarciauskas-bgse) [inline s3-discovery/Dockerfile:18]: Just an FYI that I created an issue to explore if we can de-dupe some of these Dockerfiles, https://github.com/NASA-IMPACT/cloud-optimized-data-pipelines/issues/69 feel welcome to add any ideas or thoughts you have there about de-duplication

File: s3-discovery/handler.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    files_objs = []
+    for f in filenames:
+        files_objs.append(
+            {
+                # Remove trailing back slash used for prefixing
+                "collection": event["prefix"][:-1],
+                "s3_filename": f's3://{event["bucket"]}/{f}',
+                "datetime_regex": {
+                    "regex": "^(.*?)(_)([0-9][0-9][0-9][0-9])(.tif)$",
```
Comment (abarciauskas-bgse) [inline s3-discovery/handler.py:38]: Is this regex only going to work for files that end in `.tif`? there may be a case in the future where we use this to discover files of another type in S3 for downstream processing.

**Comment by abarciauskas-bgse** [comment]
Thanks @jvntf for these contributions, mostly they look great to me just have some forward thinking ideas:

>I have defined the second state machine within the same cdk_stack.py file that was already existing, which also required creating different tasks (only by name) that use the same lambdas. This all seemed quite verbose to me, but I wasn't sure what convention is with this-- I had tried breaking it up differently but ran into issues / seemed to be a misunderstanding of how the deploy was working so I rolled back.

The way it is seems ok for now (with the exception that I don't think we need another rule target for this workflow, and actually can probably disable the existing rule / event target).

... In the future I think what we will want is to enable users to configure one or more state machines using a configuration file and a function to generate the CDK code for each state machine. A configuration file (probably written in yaml) will say "hey I want 2 workflows, one for XX collection that does cmr discovery -> stac generation and another for YY collection that does s3 -> cog -> stac generation" and that produces 2 CDK state machines. Does that make sense to you?

>I inserted a collection into the database for OMSO2PCA containing the above metadata, just to have something there so i could test the rest of the workflow

This is fine for now, minor thing but I think the second item for temporal interval I believe should be 2021 for this collection. This matters more for the final collection metadata record.

>Getting this to work requires a whole bunch of AWS config, please let me know if this is clear enough from the docs I added, or if more info is needed

There should be CDK options for most of the configuration you had to do by hand. If you still can, it would be great to have documented anything you did by hand and create a ticket to add it programmatically.

**Review by abarciauskas-bgse — APPROVED** [review]

File: cogify/README.md (hunk 1 above)
Comment (jvntf) [reply cogify/README.md:48]: i was wondering about this, ill do some research and add an issue outlining what needs to be done

File: s3-discovery/handler.py (hunk 5 above)
Comment (jvntf) [reply s3-discovery/handler.py]: good point, i updated the regex to depend on the filetype. realistically, we would will want / need to pass the regex to the discovery task as well if its chained to the stac gen task.

**Pull request merged**